### DISCUSSION TITLE
How to use custom preconditioner?

### DISCUSSION THREAD

**Discussion by andrewlhicks** [firedrakeproject/firedrake#2533] [Firedrake support] [answered]
Hello all,

I am currently working on using a custom matrix as a preconditioner for the PDE I wish to solve. I have read up a bit on Firedrake's AssembledPC class, but it is not clear from the documentation if this is what I am looking for or how I would implement it into the solve() parameters.

In other words, I just want to be able to use a simple preconditioner matrix that I specify myself. Is there an easy way to implement this?

Thanks,
Andrew

**Comment by wence-** [comment, marked as answer]
> Hello all,
> 
> I am currently working on using a custom matrix as a preconditioner for the PDE I wish to solve. I have read up a bit on Firedrake's AssembledPC class, but it is not clear from the documentation if this is what I am looking for or how I would implement it into the solve() parameters.
> 
> In other words, I just want to be able to use a simple preconditioner matrix that I specify myself. Is there an easy way to implement this?

To select a Python preconditioner you would say something like (let's say you want `AssembledPC`):

```
parameters={
  "pc_type": "python",
  "pc_python_type": "firedrake.AssembledPC",
  "assembled": {"pc_type": "lu"} # configure the PC on the assembled operator inside.
}
```

If what you want to do is supply an approximate Jacobian for your solve, there are two ways to do it. You can _either_ provide a bilinear form that defines your simple preconditioning matrix in the solve call. If you are using the `solve` interface then `solve(F == 0, Jp=your_simplified_bilinear_form)`.

The more extensible method is to implement a subclass of `AuxiliaryOperatorPC` and then select that in your solver options:

```
# Assume this is in your run-script file.

class MyPC(firedrake.AuxiliaryOperatorPC):
    _prefix = "aux" # this prefix is used in the options dictionary to configure the interior assembled solver
    def form(self, test, trial):
         # you get given test and trial functions for the space and should return a 2-tuple 
         # of a bilinear form, and any Dirichlet bcs you want to apply
         # Here a mass matrix with no bcs
         return inner(test, trial)*dx, ()

...
solver_parameters = {
  "pc_type": "python",
  "pc_python_type": "__main__.MyPC",
  # configure the sub-solver
  "aux": {"pc_type": "lu"}
}
```
